### DISCUSSION TITLE
Proposal: Introduce requirePowerOfTwo Utility Method

### DISCUSSION THREAD

**Discussion by ata-nas** [hiero-ledger/hiero-block-node#266] [Ideas]
## Proposal
We currently have an existing check if an `int` is a power of two. It uses a bitwise operation to do so:
```java
        if ((ringBufferSize & (ringBufferSize - 1)) != 0) {
            throw new IllegalArgumentException("Ring buffer size must be a power of 2");
        }
```
In my opinion this check is complex enough to become a method in a new utility class.
The logic is generic and is not coupled to any project.
It is exportable and reusable globally.
We have it already in a couple of places.

## To Discuss
- Do you agree that this could be added into the `common` module?
- If we agree to add, what should we call the new class, maybe `MathUtils` or create a class `Preconditions`?
- if we call the new class `Preconditions`, should we move also the `requireNotBlank` method from the string utils there for consistency?

@mattp-swirldslabs @AlfredoG87 @jsync-swirlds @georgi-l95 What do you guys think?

**Comment by jsync-swirlds** [comment]
Adding a `Preconditions` or `Validations` class makes sense, as does moving the `requireNotBlank` method.
I'm not sure the proposed power of two implementation is completely correct, however; it probably needs some edge guards (e.g. value >= 0, value < MAX_INT, etc...).

**Reply by ata-nas** [reply]
@jsync-swirlds noted, thanks for the comment!
Would you generally agree to add the proposed (with additional edge case guards) power for 2 check in a `Preconditions` class, alongside with moving the `requireNotBlank` string method there, in a commit that would be part of #252 or #253 since they will be done now after #251 is now available? If not we can leave for later? But keep in mind that the above check is already present from past commits. Let me know, thanks! :)

**Reply by jsync-swirlds** [reply]
I think if we have a total of 4 votes positive on the proposal and no vetoes (we probably need to define a process around this) then we should have no concerns about going forward.
Where the proposed change is made isn't terribly important; any relevant PR is good.

**Reply by ata-nas** [reply]
@jsync-swirlds that seems quite reasonable, do you have any proposals as to how we should establish the voting process and how it would look like so we can start using it? Thanks.

**Reply by jsync-swirlds** [reply]
Voting is built into discussions.
The up arrow in the lower left corner of the discussion description (and presented in the list of discussions) is an "upvote" button.

Veto would probably require a comment on the discussion explaining the reasoning.

**Reply by mattp-swirldslabs** [reply]
If we're going to make this more generic then maybe something like:
```
public static boolean isPowerOfTwo(int value) {
   return ((value & (value - 1)) != 0);
}
```

**Reply by jsync-swirlds** [reply]
The `require` method is intended to mirror the `Objects.requireNonNull` method.  I agree in principal, however.
Perhaps the following combination is acceptable:
```java
public static boolean isPowerOfTwo(int value) {
   return ((value & (value - 1)) != 0);
}

public static void requirePowerOfTwo(int value) {
   if(!isPowerOfTwo(value)) throw new IllegalArgumentException("Argument must be a power of 2.");
}

```

**Reply by ata-nas** [reply]
@jsync-swirlds was going to propose that we have both as well, I think we should have both in general. That being said, as we discussed above, I think the `Preconditions` class should be also created and the `require` methods of the non-blank strings and this new check should go there. As far as the boolean method, I think should be added at a `MathUtilities` or similar class.

**Reply by jsync-swirlds** [reply]
Both methods makes sense in a preconditions class too; one for "hard" preconditions and another for "soft" preconditions.

**Reply by ata-nas** [reply]
@jsync-swirlds In that spirit, I would also propose to do the same for the strings:

```java
// inside StringUtilities.java
public static boolean isBlank(final String toCheck) {
   return Objects.isNull(toCheck) || toCheck.isBlank();
 // shorthand for the null check as well, saves an OR every time we need to check if a string is blank
}

// inside Preconditions.java
public static void requireNotBlank(final String toCheck) {
   if(!isBlank(toCheck)) throw new IllegalArgumentException("Argument must not be blank!");
// or implemented like that ... `if Objects.requireNonNull().isBlank() throw...` as it is implemented now if you think is more appropriate that way 
}
```

**Reply by ata-nas** [reply]
@mattp-swirldslabs @jsync-swirlds since we have 4 upvotes, I will close this discussion and the proposed changes shall be implemented as discussed. See #305.
